pkg/controller/serviceaccount/legacy_serviceaccount_token_cleaner.go

```diff
@@ -0,0 +1,274 @@
+type LegacySATokenCleaner struct {
+	client clientset.Interface
+
+	saLister         listersv1.ServiceAccountLister
+	saInformerSynced cache.InformerSynced
+
+	secretLister         listersv1.SecretLister
+	secretInformerSynced cache.InformerSynced
+
+	podLister         listersv1.PodLister
+	podInformerSynced cache.InformerSynced
+
+	configMapLister         listersv1.ConfigMapLister
+	configMapInformerSynced cache.InformerSynced
+
+	legacySATokenCleanUpInterval time.Duration
+	legacySATokenCleanUpPeriod   time.Duration
+}
+
+// NewLegacySATokenCleaner returns a new *NewLegacySATokenCleaner.
+func NewLegacySATokenCleaner(saInformer coreinformers.ServiceAccountInformer, secretInformer coreinformers.SecretInformer, podInformer coreinformers.PodInformer, configMapInformer coreinformers.ConfigMapInformer, cl clientset.Interface, options LegacySATokenCleanerOptions) (*LegacySATokenCleaner, error) {
+	tc := &LegacySATokenCleaner{
+		client:                       cl,
+		saLister:                     saInformer.Lister(),
+		saInformerSynced:             saInformer.Informer().HasSynced,
+		secretLister:                 secretInformer.Lister(),
+		secretInformerSynced:         secretInformer.Informer().HasSynced,
+		configMapLister:              configMapInformer.Lister(),
+		configMapInformerSynced:      configMapInformer.Informer().HasSynced,
+		podLister:                    podInformer.Lister(),
+		podInformerSynced:            podInformer.Informer().HasSynced,
+		legacySATokenCleanUpPeriod:   options.LegacySATokenCleanUpPeriod,
+		legacySATokenCleanUpInterval: options.LegacySATokenCleanUpInterval,
+	}
+
+	return tc, nil
+}
+
+func (tc *LegacySATokenCleaner) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting legacy service account token cleaner controller")
+	defer logger.Info("Shutting down legacy service account token cleaner controller")
+
+	if !cache.WaitForNamedCacheSync("legacy-service-account-token-cleaner", ctx.Done(), tc.saInformerSynced, tc.secretInformerSynced, tc.configMapInformerSynced, tc.podInformerSynced) {
+		return
+	}
+
+	go wait.UntilWithContext(ctx, tc.evaluateSATokens, tc.legacySATokenCleanUpInterval)
+
+	<-ctx.Done()
+}
+
+func (tc *LegacySATokenCleaner) evaluateSATokens(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	secretList, err := tc.secretLister.Secrets(metav1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "Getting secret list from secret cache")
+		return
+	}
+	for _, secret := range secretList {
+		if secret.Type != v1.SecretTypeServiceAccountToken {
+			continue
+		}
+		sa, saErr := tc.getServiceAccount(secret)
+		switch {
+		case saErr != nil:
+			logger.Error(saErr, "Getting service account")
+		case sa == nil:
+			logger.V(4).Info("Service account does not exist", "secret", klog.KRef(secret.Namespace, secret.Name))
```

Comment:
these would be easier to follow as unnested `if ... { continue }` blocks

```go
if saErr != nil {
  logger.Error(saErr, "Getting service account")
  continue
}
if sa == nil || !hasSecretReference(sa, secret.Name) {
  // can't determine if this is an auto-generated token
  continue
}

mountedSecretNames, err := getMountedSecretNames(secret.Namespace)
if err != nil {
  logger.ErrorS(saErr, "resolving mounted secrets", "namespace", secret.Namespace)
  continue
}
if mountedSecretNames.Has(secret.Name) {
  // still used by pods
  continue
}

logger.Info("Auto-generated service account token is not used...
if err := tc.client.CoreV1().Secrets(secret.Namespace).Delete(ctx, secret.Name, metav1.DeleteOptions{Preconditions: &metav1.Precondition{ResourceVersion: &secret.ResourceVersion}}); err != nil && !apierrors.IsConflict(err) && !apierrors.IsNotFound(err) {
...
}
```

`getMountedSecretNames(namespace string) (sets.String, error)` can be a locally defined function in `evaluateSATokens` that resolves the names of secrets used by pods in the given namespace once per invocation of `evaluateSATokens`

put a resourceVersion precondition on the secret delete as well, so if anything touched it since we observed it, we don't delete without considering the latest state of the object

ignore notfound and conflict errors coming back from the delete call... we don't care if something we tried to remove is already removed, and a conflict error will get retried next sync